ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/AddMetadata.java

```diff
@@ -4,45 +4,173 @@
+    final String namespace = attributes.get("document_namespace");
+    Map<String, Object> metadata = new HashMap<>();
+    metadata.put("geo", geoFromAttributes(attributes));
+    metadata.put("user_agent", userAgentFromAttributes(attributes));
+    metadata.put("header", headersFromAttributes(attributes));
+    if ("telemetry".equals(namespace)) {
+      metadata.put("uri", uriFromAttributes(attributes));
+    }
+    Map<String, Object> payload = new HashMap<>();
+    payload.put("metadata", metadata);
+    TOP_LEVEL_STRING_FIELDS.forEach(name -> Optional //
+        .ofNullable(attributes.get(name)) //
+        .ifPresent(value -> payload.put(name, value)));
+    TOP_LEVEL_INT_FIELDS.forEach(name -> Optional //
+        .ofNullable(attributes.get(name)) //
+        .flatMap(value -> {
+          try {
+            return Optional.of(Integer.parseInt(value));
+          } catch (NumberFormatException e) {
+            return Optional.empty();
+          }
+        }) //
+        .ifPresent(value -> payload.put(name, value)));
+    return payload;
+  }
+
+  static void stripPayloadMetadataToAttributes(Map<String, String> attributes,
+      Map<String, Object> payload) {
+    Optional.ofNullable(payload) //
+        .map(p -> p.remove("metadata")).filter(Map.class::isInstance) //
+        .ifPresent(m -> {
+          Map<String, Object> metadata = (Map<String, Object>) m;
+          putGeoAttributes(attributes, metadata);
+          putUserAgentAttributes(attributes, metadata);
+          putHeaderAttributes(attributes, metadata);
+          putUriAttributes(attributes, metadata);
+        });
+    TOP_LEVEL_STRING_FIELDS.forEach(name -> Optional //
+        .ofNullable(payload).map(p -> p.remove(name))
+        .ifPresent(value -> attributes.put(name, value.toString())));
+    TOP_LEVEL_INT_FIELDS.forEach(name -> Optional //
+        .ofNullable(payload).map(p -> p.remove(name))
+        .ifPresent(value -> attributes.put(name, value.toString())));
+  }
+
+  static Map<String, Object> geoFromAttributes(Map<String, String> attributes) {
+    HashMap<String, Object> geo = new HashMap<>();
+    attributes.keySet().stream() //
+        .filter(k -> k.startsWith("geo_")) //
+        .forEach(k -> geo.put(k.replaceFirst("geo_", ""), attributes.get(k)));
+    return geo;
+  }
+
+  static void putGeoAttributes(Map<String, String> attributes, Map<String, Object> metadata) {
+    Optional.ofNullable(metadata).map(m -> m.get("geo")) //
+        .filter(Map.class::isInstance) //
+        .map(m -> ((Map<String, Object>) m).entrySet()) //
+        .orElse(ImmutableSet.of()) //
+        .forEach(entry -> attributes.put("geo_" + entry.getKey(), entry.getValue().toString()));
+  }
+
+  static Map<String, Object> userAgentFromAttributes(Map<String, String> attributes) {
+    HashMap<String, Object> userAgent = new HashMap<>();
+    attributes.keySet().stream() //
+        .filter(k -> k.startsWith("user_agent_")) //
```

Comment:
`"user_agent"` and `"geo"` are referenced here enough times that it seems appropriate to pull them into constants so that the compiler can add some safety around spelling